README.txt
SOUND _ FALL

-------------------------------
Created by Chris Thompson
http://www.stipes.org/

Weekly Challenge #1: NON_VISUAL
V1.0 - 8 March 2010
-------------------------------

## WHAT YOU'LL NEED:

 - Python (http://www.python.org/, tested in Version 2.6)
 - Pygame (http://www.pygame.org/)
 - Headphones

## HOW TO PLAY:

Navigate to where you unpacked the game files with your terminal of choice.
To run the game from the folder, type:

  python soundfall.py

This will open a window with instructions. Some output may be printed to the
terminal as well.

The arrow keys move the sound block around the grid. The space bar places the
sound block and starts a new sound. ESCAPE exits the game.

(to cheat, or to make sure you're on the right course, press ENTER to 
print the grid to the terminal.)

Note: Pygame is not included in this distribution. Due to the use of the 
terminal, I am not able to convert this to a .EXE for Windows users.

## LICENSE:

Sound _ Fall
Copyright (C) 2010  Chris Thompson

This game and its included resources are available under the 
GNU PUblic License V3. See LICENSE.txt for more information.

Basically, you may use this however you want, as long as you make your
addittions open. I particularly hope that if you use it, you do something cool.
If you do, tell me about it! I'd also be cool with a small attribution line.

If you want to use this in a different way (non-GPL), get in touch.

## ACKNOWLEDGEMENTS:

Making this game work was harder than I thought. I jumped between OpenAL, CLAM,
cSound, and many other systems. Most were ridiculously hard to get working
with Python in Linux (CLAM, SuperCollider, PyOpenAL, etc.) or not quite the
right tool for the job (cSound, which I still love). I ended up building a 
custom panning system that sits on top of Pygame.Mixer. You can find out more
about PyGame at pygame.org.

Sound samples were created in Audacity, and images were created in GIMP.


PyOpenAL-0.1.6/CHANGES

* March 2003 : PyOpenAL 0.1
  * First release

* March 2003 : PyOpenAL 0.1.1
  * openal4soya now automagically compute (by derivation) the source speed.
  * openal4soya can now load sound asynchronously
  * fixed a bug in the test.py script (openal replaced by pyopenal)
  
* April 2003 : PyOpenAL 0.1.2
  * PyOpenAL is now licensed under the LGPL (instead of GPL);
    this choice sounds because OpenAL is LGPL'ed too.
  * sound played asynchronously are now canceled if stop() is called
    before the sound starts

* December 2003 : PyOpenAL 0.1.3
  * minor bugfixes

* December 2003 : PyOpenAL 0.1.4
  * Moved openal4soya into soya
  * Bugfixes:
    * Remove "home_page" keyword in setup.py, which cause troubles
    with old python

* July 2005 : PyOpenAL 0.1.5
  * Now compile with the latest OpenAL CVS (thanks Daja77 !)

* April 2006 : PyOpenAL 0.1.6
  * Update for new OpenAL with libopenal and libalut (thanks Duck !)




soundfall_al.py
#! /bin/python

# yay

import os, os.path, sys, time
#import pprint
import pyopenal
import pygame
from pygame.locals import *

#pp = pprint.PrettyPrinter(indent=2)

pyopenal.init(None)
l = pyopenal.Listener(44100)
#             x    y    z
l.position = (0.0, 0.0, 0.0) # at origin
l.at       = (0.0, 0.0, -1.0) # facing z = -1 (into screen)
l.up       = (0.0, 1.0, 0.0) # y=1 is our "up"

# we build an array of buffers to house the vertical "positions"
buf = [
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "0.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "1.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "2.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "3.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "4.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "5.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "6.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "7.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "8.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "9.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "10.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "11.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "12.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "13.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "14.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "15.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "16.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "17.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "18.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "19.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "20.wav")),
pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "21.wav"))  ]

colbuf = pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sounds", "collission.wav"))

class Piece:
    "Maintains a piece."
    global buf, colbuf
    def __init__(self, xstart=6, ystart=21, dist=2):
        # position on grid
        self.xpos = xstart
        self.ypos = ystart
        self.zpos = dist
        self.source = pyopenal.Source()
        # position in space
        self.source.position = ((xstart - 6) * 0.2 , ystart * 0.1, dist)
        self.source.velocity = (0.0, 0.0, 0.0)
        self.source.buffer = buf[ystart]
        self.source.looping = 0
        self.source.play()
        
        # collission source
        self.col = pyopenal.Source()
        self.col.position = self.source.position
        self.col.buffer = colbuf
        self.source.looping = 0
    
    def move(self, dir, board):
        # Check the move, update on grid
        # Need to add collission checking with prior pieces
        if dir == 'l':
            print self.xpos
            print self.ypos
            if self.xpos <= 0: self.col.play()
            elif board.grid[self.xpos - 1][self.ypos]:
                self.col.play()
                return
            elif self.xpos > 0: self.xpos -= 1
        
        if dir == 'r':
            print self.xpos
            print self.ypos
            if self.xpos >= 11: self.col.play()
            elif board.grid[self.xpos + 1][self.ypos]: self.col.play()
            elif self.xpos < 11: self.xpos += 1

        
        if dir == 'u':
            print self.xpos
            print self.ypos
            if self.ypos >= 21: self.col.play()
            elif board.grid[self.xpos][self.ypos + 1]: self.col.play()
            elif self.ypos < 21: self.ypos += 1
        
        if dir == 'd':
            print self.xpos
            print self.ypos
            if self.ypos <= 0: self.col.play()
            elif board.grid[self.xpos][self.ypos - 1]: self.col.play()
            elif self.ypos > 0: self.ypos -= 1
        
        # Change buffer to change the "height" of the sound
        self.source.buffer = buf[self.ypos]
        # Update the 3d position
        self.source.position = ((self.xpos - 6) * 0.2, self.ypos * 0.1, self.zpos)
        
        self.source.play()
        
        self.col.position = self.source.position
        
        time.sleep(2)
    
    def place(self):
        self.source.stop()

class Board:
    "Maintain state of top-down grid."
    def __init__(self, width=12, height=22):
        # grid[x][y]
        self.grid = [[0] * height for i in xrange(width)]
    
    def toggleSquare(self, x, y):
        self.grid[x][21 - y] = not self.grid[x][21 - y]
        
    def display(self, piece):
#        gridCopy = deepcopy(self.grid)
#        gridCopy[piece.xpos, piece.ypos] = True
#        for row, i in enumerate(gridCopy):
#            for square, j in enumerate(row):
#                if square: gridCopy[i][j] = 'X'
#                else: gridCopy[i][j] = ' '
        print ''
        for i,row in enumerate(zip(*self.grid)):
            print '| ',
            for j,square in enumerate(row):
                if square: print ' X ',
                elif (21-i == piece.ypos and j == piece.xpos): print ' X ',
                else: print '   ',
            print ' |'
        print ''

def main():
    # Set up for game
    
    pygame.init()
    
    # Do I need a screen to catch input? Fine.
    
    screen = pygame.display.set_mode((480,300))
    pygame.display.set_caption('SOUND_FALL')
    pygame.mouse.set_visible(0)
    # write some instructions to this screen?
    
    
    # Play an intro bit? Verbal tutorial with training for
    # TOP and BOTTOM, left/right, etc?
    
    board = Board()
    piece = Piece()
    
    # make sure that the piece CAN enter?
    #board.enterPiece(piece)
    
    while 1:
        # GAME LOOP
        
        # get input -- NOTE: we only listen to keydowns currently
        for event in pygame.event.get():
            if event.type == QUIT:
                pygame.quit()
                return
            if event.type == KEYDOWN:
                if event.key == K_ESCAPE:
                    pygame.quit()
                    return
                elif event.key == K_UP:
                    piece.move('u',board)
                    
                elif event.key == K_DOWN:
                    piece.move('d',board)
                    
                elif event.key == K_LEFT:
                    piece.move('l',board)
                    
                elif event.key == K_RIGHT:
                    piece.move('r',board)
                    
                elif event.key == K_SPACE:
                    piece.place()
                    board.toggleSquare(piece.xpos, piece.ypos)
                    piece = Piece()
                    
                elif event.key == K_RETURN:
                    board.display(piece)
        
        
    
    # End of game procedure
    
if __name__ == '__main__': main()


PyOpenAL-0.1.6/__init__.py
# PyOpenAl
# Copyright (C) 2003 Jean-Baptiste LAMY -- jiba@tuxfamily.org
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

"""pyopenal

OpenAL binding for Python.

File formats supported:
 - Wave
 - Ogg Vorbis (needs PyOgg and PyVorbis)

The binding does not include yet OpenAL extensions.

See test.py for an example."""

from _openal import *
import _openal

from cStringIO import StringIO

import atexit
atexit.register(quit)

_INITED = 0
def init(arg = "'( ( devices '( native esd sdl alsa arts null ) ) )"):
  """init(arg = "'( ( devices '( native esd sdl alsa arts null ) ) )"

Inits OpenAL. ARG can be used to select which device to use."""
  global _INITED
  if not _INITED:
    _openal.init(arg)
    _INITED = 1

class Listener(_openal._Listener):
  """Listener

A listener that listens sounds."""
  pass

class Source(_openal._Source):
  """Source

A source that produces sound."""
  pass

Buffer = _openal._Buffer
# class Buffer(_openal._Buffer):
#   def __init__(self, format = None, data = None, freq = 0):
#     _openal._Buffer.__init__(self)
#     if not format is None:
#       self.set_data(format, data, freq)


def wave_data(filename):
  """wave_data(filename) -> format, data, freq

Get the data from a Wave file, in a format suitable for Buffer.set_data."""
  
  import wave
  wav  = wave.open(filename)
  data = wav.readframes(wav.getnframes())
  
  if wav.getnchannels() == 2:
    if wav.getsampwidth() == 2: return AL_FORMAT_STEREO16, data, wav.getframerate()
    else:                       return AL_FORMAT_STEREO8 , data, wav.getframerate()
  else:
    if wav.getsampwidth() == 2: return AL_FORMAT_MONO16  , data, wav.getframerate()
    else:                       return AL_FORMAT_MONO8   , data, wav.getframerate()
    
def ogg_vorbis_data(filename):
  """ogg_vorbis_data(filename) -> format, data, freq

Get the data from an Ogg Vorbis file, in a format suitable for Buffer.set_data.
Require PyOgg and PyVorbis"""
  
  import ogg.vorbis
  vorb = ogg.vorbis.VorbisFile(filename)
  info = vorb.info()
  
  buffer = Buffer()
  
  bytes = 1
  data  = StringIO()
  while bytes:
    buff, bytes, bit = vorb.read(4096)
    data.write(buff[:bytes])
  data = data.getvalue()
  
  # Assume Ogg Vorbis file are 16 bit sounds !!!
  # XXX I'm not sure of that !!!
  
  if info.channels == 2: return AL_FORMAT_STEREO16, data, info.rate 
  else:                  return AL_FORMAT_MONO16  , data, info.rate 


def WaveBuffer(filename):
  """WaveBuffer(filename) -> Buffer

Loads a Wave file, for playing with OpenAL."""
  buffer = Buffer()
  buffer.set_data(*wave_data(filename))
  return buffer

def OggVorbisBuffer(filename):
  """OggVorbisBuffer(filename) -> Buffer

Loads an Ogg Vorbis file, for playing with OpenAL.
Require PyOgg and PyVorbis"""
  
  buffer = Buffer()
  buffer.set_data(*ogg_vorbis_data(filename))
  return buffer
  


PyOpenAL-0.1.6/demo/demo.py
import os, os.path, sys, time

import pyopenal

pyopenal.init(None)
l = pyopenal.Listener(22050)
l.position = (0.0, 0.0, 0.0)
l.at = (0.0, 0.0, 1.0)
l.up = (0.0, 1.0, 0.0)

b = pyopenal.WaveBuffer(os.path.join(os.path.dirname(sys.argv[0]), "sword.wav"))
#b = pyopenal.OggVorbisBuffer("/home/jiba/src/slune/sounds/summum_of_the_light.ogg")

s = pyopenal.Source()
s.position = (0.0, 0.0, 0.0)
s.velocity = (0.0, 0.0, 0.0)
s.buffer  = b
s.looping = 1

s.play()

# Move the source further

z = 0.0
i = 0
while i < 100:
  z += 0.5
  s.position = (0.0, 0.0, z)
  time.sleep(0.1)
  i = i + 1

print 'ok'


## device = openal.alcOpenDevice(None)
## #device = openal.alcOpenDevice("'( ( devices '( native waveout esd sdl alsa arts null ) ) )")
## context = openal.alcCreateContext (device, (openal.ALC_FREQUENCY, 22050, openal.ALC_INVALID))
## openal.alcMakeContextCurrent (context)


## openal.alListenerfv (openal.AL_POSITION, (0.0, 0.0, 0.0))
## openal.alListenerfv (openal.AL_ORIENTATION, (0.0, 0.0, 1.0, 0.0, 1.0, 0.0))
## stereo = openal.alGenBuffers (1)

## file = open('./sword.wav', 'r')
## data = file.read()
## openal.alBufferData (stereo, openal.AL_FORMAT_WAVE_EXT, data, 0)
## moving_source = openal.alGenSources (1)
     
## openal.alSource3f (moving_source, openal.AL_POSITION, 0.0, 0.0, 20.0)
## openal.alSourcefv (moving_source, openal.AL_VELOCITY, (0.0, 0.0, 0.0))
## openal.alSourcei  (moving_source, openal.AL_BUFFER, stereo)
## openal.alSourcei  (moving_source, openal.AL_LOOPING, openal.AL_FALSE)
   
## openal.alSourcePlayv (moving_source)
## #openal.alSourcei (moving_source, openal.AL_SOURCE_STATE, openal.AL_INITIAL)

## print "Using OpenAL", openal.alGetString(openal.AL_VERSION), "by", openal.alGetString(openal.AL_VENDOR)

## while(1): pass


soundfall.py
#! /bin/python

# SOUND _ FALL
# A game where you place sound. Written in Python with Pygame.
#    Copyright (C) 2010  Chris Thompson (http://www.stipes.org)

#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.

#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.

#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.

import os, os.path, sys, time, math
import pygame
from pygame.locals import *

if not pygame.mixer: print "No sound system detected."
pygame.mixer.init(44100)

# initialize sound objects
collissionSound = pygame.mixer.Sound(os.path.join("sounds","collission.wav"))
sounds = [
pygame.mixer.Sound(os.path.join("sounds","0.wav")),
pygame.mixer.Sound(os.path.join("sounds","1.wav")),
pygame.mixer.Sound(os.path.join("sounds","2.wav")),
pygame.mixer.Sound(os.path.join("sounds","3.wav")),
pygame.mixer.Sound(os.path.join("sounds","4.wav")),
pygame.mixer.Sound(os.path.join("sounds","5.wav")),
pygame.mixer.Sound(os.path.join("sounds","6.wav")),
pygame.mixer.Sound(os.path.join("sounds","7.wav")),
pygame.mixer.Sound(os.path.join("sounds","8.wav")),
pygame.mixer.Sound(os.path.join("sounds","9.wav")),
pygame.mixer.Sound(os.path.join("sounds","10.wav")),
pygame.mixer.Sound(os.path.join("sounds","11.wav")),
pygame.mixer.Sound(os.path.join("sounds","12.wav")),
pygame.mixer.Sound(os.path.join("sounds","13.wav")),
pygame.mixer.Sound(os.path.join("sounds","14.wav")),
pygame.mixer.Sound(os.path.join("sounds","15.wav")),
pygame.mixer.Sound(os.path.join("sounds","16.wav")),
pygame.mixer.Sound(os.path.join("sounds","17.wav")),
pygame.mixer.Sound(os.path.join("sounds","18.wav")),
pygame.mixer.Sound(os.path.join("sounds","19.wav")),
pygame.mixer.Sound(os.path.join("sounds","20.wav")),
pygame.mixer.Sound(os.path.join("sounds","21.wav")) ]

def pan(x=0, y=0, d=2):
    "Basic 2D panning with 3D attenuation."
    
    # we use a constant power of 1 for the pan, so strong L/R differentiation
    left = .5
    right = .5
    
    # basic conversion of grid to space for the pan using lower part of cosine
    # means that at x=12 -> right=1 left = 0
    # and at x=0 -> right=0 left=1
    diff = 0
    # direction to pan
    if (x - 6) < 0:
        diff = math.cos( (x-6)*math.pi/12 - math.pi/2) * 0.5
        left -= diff
        right += diff
    elif (x - 6) > 0:
        diff = math.cos( (x-6)*math.pi/12 + math.pi/2) * 0.5
        left += diff
        right -= diff
    
    return left, right

class Piece:
    "Maintains a piece."
    global sounds, collissionSound
    def __init__(self, xstart=6, ystart=21, dist=2):
        # position on grid
        self.xpos = xstart
        self.ypos = ystart
        self.zpos = dist
        self.source = pygame.mixer.Channel(0)
        self.collission = pygame.mixer.Channel(1)
        # position in space
        #((xstart - 6) * 0.2 , ystart * 0.1, dist)
        left, right = pan(xstart, dist)
        self.source.set_volume(left, right)
        self.collission.set_volume(left,right)
        self.source.play(sounds[ystart], 0) # start, don't loop
    
    def move(self, dir, board):
        # Check the move, update on grid
        # Collission detecting only works on move,
        # so a new piece can start inside an old piece
        
        self.source.stop()
        
        if dir == 'l':
            if self.xpos <= 0:
                self.collission.play(collissionSound)
            elif board.grid[self.xpos - 1][21 - self.ypos]:
                self.collission.play(collissionSound)
            elif self.xpos > 0:
                self.xpos -= 1
        
        if dir == 'r':
            if self.xpos >= 12:
                self.collission.play(collissionSound)
            elif board.grid[self.xpos + 1][21 - self.ypos]:
                self.collission.play(collissionSound)
            elif self.xpos < 12:
                self.xpos += 1
        
        if dir == 'u':
            if self.ypos >= 21:
                self.collission.play(collissionSound)
            elif board.grid[self.xpos][21 - (self.ypos + 1)]:
                self.collission.play(collissionSound)
            elif self.ypos < 21:
                self.ypos += 1
        
        if dir == 'd':
            if self.ypos <= 0:
                self.collission.play(collissionSound)
            elif board.grid[self.xpos][21 - (self.ypos - 1)]:
                self.collission.play(collissionSound)
            elif self.ypos > 0:
                self.ypos -= 1
        
        self.source.stop()
        
        # Play the new "height" of the sound
        self.source.play(sounds[self.ypos], 0)
        
        # Update the panning
        left, right = pan(self.xpos, self.ypos, self.zpos)
        self.source.set_volume(left, right)
        self.collission.set_volume(left, right)
        
        time.sleep(0.1)
    
    def place(self):
        self.source.stop()

class Board:
    "Maintain state of top-down grid."
    def __init__(self, width=13, height=22):
        # grid[x][y]
        self.grid = [[0] * height for i in xrange(width)]
    
    def toggleSquare(self, x, y):
        self.grid[x][21 - y] = not self.grid[x][21 - y]
        
    def display(self, piece):
        print ''
        for i,row in enumerate(zip(*self.grid)):
            print '| ',
            for j,square in enumerate(row):
                if square: print ' X ',
                elif (21-i == piece.ypos and j == piece.xpos): print ' X ',
                else: print '   ',
            print ' |'
        print ''

def main():
    pygame.init()
    
    # Set up screen with instructions
    screen = pygame.display.set_mode((480,300))
    pygame.display.set_caption('SOUND_FALL')
    pygame.mouse.set_visible(0)
    title = pygame.image.load("title.png")
    title = title.convert()
    screen.blit(title, (0,0))
    pygame.display.flip()
    
    board = Board()
    piece = Piece()
    
    while 1:
        # GAME LOOP
        
        # get input -- NOTE: we only listen to keydowns currently
        for event in pygame.event.get():
            if event.type == QUIT:
                pygame.quit()
                return
            if event.type == KEYDOWN:
                if event.key == K_ESCAPE:
                    pygame.quit()
                    return
                elif event.key == K_UP:
                    piece.move('u',board)
                    
                elif event.key == K_DOWN:
                    piece.move('d',board)
                    
                elif event.key == K_LEFT:
                    piece.move('l',board)
                    
                elif event.key == K_RIGHT:
                    piece.move('r',board)
                    
                elif event.key == K_SPACE:
                    piece.place()
                    board.toggleSquare(piece.xpos, piece.ypos)
                    piece = Piece()
                    
                elif event.key == K_RETURN:
                    board.display(piece)
    pygame.quit()
    
if __name__ == '__main__': main()
